lib/bad_list_subscriptions_migrator.rb

```diff
@@ -0,0 +1,44 @@
+class BadListSubscriptionsMigrator
+  attr_reader :prefix
+
+  def initialize(prefix)
+    @prefix = prefix
+  end
+
+  def process_all_lists
+    subscriber_list_urls.each do |url|
+      candidate_lists = SubscriberList.where(url:)
+      move_users_from_bad_lists_to_good_list(candidate_lists)
+    end
+  end
+
+private
+
+  def subscriber_list_urls
+    @subscriber_list_urls ||= subscriber_lists.pluck(:url).uniq
+  end
+
+  def subscriber_lists
+    @subscriber_lists ||= SubscriberList.where("url LIKE ?", "%/#{prefix}/%")
+  end
+
+  def move_users_from_bad_lists_to_good_list(lists)
+    valid_lists = lists.select { |list| valid_links_based_list?(list) }
+    bad_lists = lists - valid_lists
+
+    if valid_lists.count == 1
```

Comment:
Just a check - do we have any examples where there is no valid list, but there is a bad list? I imagine it's unlikely, but could theoretically have happened if all the signups were when the bug was in place, and none before / after?

Reply:
Yep good shout, thanks 👍

Reply:
And I guess also are there any with a valid lists count of 2?

Reply:
Joining up the comments... Keith looked into this on staging (comment [here](https://docs.google.com/document/d/1Ci61Neb8sJqGn31DUNuhVS-pT1-RJnM6gtPpYNe8wi8/edit?disco=AAAAyi_ScYQ))

Reply:
Keith's comment:
```
I checked on Staging, and out of 37156 subscriber lists, there are 3 urls which have 3 subscriber lists: 

- /government/organisations/hm-passport-office
- /government/organisations/groceries-code-adjudicator
- /government/organisations/building-regulations-advisory-committee.

They all follow the same pattern:
1st list is for the slug "all-types-of-document-about-<whatever>", and has a links digest, 
2nd is for the slug "<whatever>" and has a links digest, 
3rd is for the slug "<whatever>-<8-digit-hex-number>" and has a content_id)
```
So let's look at the 3 subscriber lists with the url "/government/organisations/hm-passport-office"

The two good lists both have the same links (and therefore the same links digest). They both have a similar number of active subscriptions. And fetching the most recently created subscription for these lists reveals both are actively being signed up for. 

So this is what I'm looking into...

How has this situation been created?
Does it matter?
If it matters --> leave this check in, and manually handle the exceptions after running the task.
If it doesn't matter --> remove the check

Reply:
- How has this situation been created?

When Email Alert API creates a new subscriber list, the [slug is formed](https://github.com/alphagov/email-alert-api/blob/c445e17ee1598c40d6fca7a985ddda19eae5dec5/app/services/create_subscriber_list_service.rb#L34) from the title parameter [sent from email alert frontend](https://github.com/alphagov/email-alert-frontend/blob/main/app/services/subscriber_list_params/generate_links_based_list_params_service.rb#L10). So it's possible that at some point the code creating slugs in email alert api, or the code populating the title parameter in email alert frontend, has changed? I can't find evidence of this.

- Does it matter?

When a user clicks to sign up to a links based email subscription, email alert frontend asks email alert api to [find or create a subscriber list with the provided params](https://github.com/alphagov/email-alert-frontend/blob/757c60919a9fc11452d642876e6c54b5f96d73fb/app/controllers/content_item_signups_controller.rb#LL36C25-L36C25).

In email alert api, the [create_subscriber_list_service](https://github.com/alphagov/email-alert-api/blob/c445e17ee1598c40d6fca7a985ddda19eae5dec5/app/services/create_subscriber_list_service.rb#L13) uses the [find_exact_query](https://github.com/alphagov/email-alert-api/blob/main/app/queries/find_exact_query.rb#L23) class to match the provided params to existing subscriber lists in its database. Because links are present, it will [match on links_digest](https://github.com/alphagov/email-alert-api/blob/c445e17ee1598c40d6fca7a985ddda19eae5dec5/app/queries/find_exact_query.rb#L57) first. I think [this](https://stackoverflow.com/questions/42855244/how-it-works-find-by-in-rails) is what rails does under the hood (looking for a better ref) ie it returns the first record that matches. So in our case it will sometimes return the "all-types-of-document-about-<whatever>" list, and sometimes it will return "<whatever>" list.  So users are being subscribed to either of these. 

When a content change is sent to email alert api, it uses the SubscriberListQuery - will it match against both of these lists, or only one? To be continued...

Reply:
I believe a matched content change will match on both of these lists according to the [matched_for_notification](https://github.com/alphagov/email-alert-api/blob/c445e17ee1598c40d6fca7a985ddda19eae5dec5/app/queries/matched_for_notification.rb#L20) query. It does a `select` for all SubscriberLists that match on links. Rather than a `find_by` that returns the first match it finds.

So where does that leave us!

I can remove this check for if there is only one good list, and just use the first list. 

So TLDR:

No, it doesn't matter which list the user is subscribed to

Reply:
Further probing into this confirms that both lists are valid and either can be used as the target for migration:

```
hm_lists = SubscriberList.where(url: "/government/organisations/hm-passport-office")
hm_lists.count
=> 3

ids = hm_lists.pluck(:id)
249080 is a bad list (content id no links)
4256 and 4655 are both good lists that differ only by slug and creation/updated_at timestamps

4256 slug: "all-types-of-document-about-all-topics-by-hm-passport-office",
4655 slug: slug: "hm-passport-office",

content_change_ids_for_4655 =  MatchedContentChange.where(subscriber_list_id: 4655).pluck(:content_change_id).sort

content_change_ids_for_4256 =  MatchedContentChange.where(subscriber_list_id: 4256).pluck(:content_change_id).sort

content_change_id_id_4655 == content_change_id_id_4256
=> true
```